deeppavlov/pipeline_manager/pipeline_manager.py

```diff
@@ -0,0 +1,519 @@
+
+            if len(self.available_gpu) == 0:
+                raise ValueError("All selected GPU with numbers: ({}), are busy.".format(set(self.use_gpu)))
+            elif len(self.available_gpu) < len(self.use_gpu):
+                print("PipelineManagerWarning: 'CUDA_VISIBLE_DEVICES' = ({0}), "
+                      "but only {1} are available.".format(self.use_gpu, self.available_gpu))
+
+            self.max_num_workers = len(self.available_gpu)
+
+    @staticmethod
+    @unpack_args
+    def train_pipe(pipe: Dict, i: int, observer_: ExperimentObserver, gpu_ind: Optional[int] = None) -> None:
+        """
+        Start learning single pipeline. Observer write all info in log file.
+
+        Args:
+            pipe: config dict of pipeline
+            i:  number of pipeline
+            observer_: link to observer object
+            gpu_ind: number of gpu to use (if multiprocessing is True)
+
+        """
+        # modify project environment
+        if gpu_ind:
+            os.environ['CUDA_VISIBLE_DEVICES'] = str(gpu_ind)
+        else:
+            os.environ['CUDA_VISIBLE_DEVICES'] = ''
+
+        dataset_name = copy(pipe['dataset_reader']['data_path'].split("/")[-1])
+
+        observer_.pipe_ind = i + 1
+        observer_.pipe_conf = copy(pipe['chainer']['pipe'])
+        observer_.batch_size = copy(pipe['train'].get('batch_size', "None"))
+
+        # start pipeline time
+        pipe_start = time.time()
+
+        save_path = observer_.save_path / dataset_name / "pipe_{}".format(i + 1)
+        if not save_path.exists():
+            save_path.mkdir(parents=True)
+
+        # run pipeline train with redirected output flow
+        proc_out_path = save_path / f"out_{i + 1}.txt"
+        proc_err_path = save_path / f"err_{i + 1}.txt"
+        with proc_out_path.open("a", encoding='utf8') as out_file, proc_err_path.open("a", encoding='utf8') as err_file:
+            with redirect_stderr(err_file), redirect_stdout(out_file):
+                results = train_evaluate_model_from_config(pipe, to_train=True, to_validate=True)
+
+        # add results and pipe time to log
+        observer_.pipe_time = time.strftime('%H:%M:%S', time.gmtime(time.time() - pipe_start))
+        observer_.pipe_res = results
+
+        # update logger
+        observer_.update_log()
+
+        # save config in checkpoint folder
+        observer_.save_config(pipe, dataset_name, i + 1)
+
+    def gpu_gen(self, gpu: bool = False) -> Generator:
+        """
+        Create generator that returning tuple of args fore self.train_pipe method.
+
+        Args:
+            gpu: boolean trigger, determine to use gpu or not
+
+        """
+        if gpu:
+            for i, pipe_conf in enumerate(self.pipeline_generator()):
+                gpu_ind = i - (i // len(self.available_gpu)) * len(self.available_gpu)
+                yield (deepcopy(pipe_conf), i, self.observer, gpu_ind)
```

Comment:
If you're getting `gpu_ind` from a generator and are running in parallel, gpus can free up not in the same order that ids are provided